### DISCUSSION TITLE
Unbound and Redis

### DISCUSSION THREAD

**Discussion by Dynamic5912** [MichaIng/DietPi#7210] [Q&A]
Does Unbound install via DietPi with the requirements/dependencies for utilising Redis?

I understand that Unbound has to be installed with **cachedb** and some other dependencies to work with Redis. 

Thanks!

**Comment by Joulinar** [comment]
We install the default Debian package, which doesn't support Redis.

**Reply by Dynamic5912** [reply]
OK. That's a shame.

**Reply by Joulinar** [reply]
Why is that a shame? Because we install standard Debian packages? Everyone is welcome to contribute to our project and provide packages containing such features. Unfortunately DietPi has only one developer and therefore it is almost impossible to maintain and keep every software package up-to-date. Therefore, like almost every Debian destro, we use the global Debian packages repository.

**Reply by MichaIng** [reply]
I am actually surprised as well, that the Debian package does not support Redis for caching, if Unbound generally does support it. Usually, Debian builds software with all options enabled. I see no reason to have Redis caching option disabled 🤔.

**Reply by MichaIng** [reply]
And it does in fact compile with `--enable-cachedb --with-libhiredis`: https://buildd.debian.org/status/fetch.php?pkg=unbound&arch=amd64&ver=1.20.0-1%2Bb1&stamp=1719260287&raw=0

And here is the manpage for the Bookworm package, which contains all instructions to configure Unbound for using Redis as cachdb backaned: https://manpages.debian.org/unbound.conf

**Reply by Dynamic5912** [reply]
I installed unbound via DietPi-Software and cannot see that it has cachedb included:

_dietpi@dietpi:~$ sudo unbound -V
Version 1.17.1

Configure line: --build=aarch64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir=${prefix}/lib/aarch64-linux-gnu --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking --with-pythonmodule --with-pyunbound --enable-subnet --enable-dnstap --enable-systemd --with-libnghttp2 --with-chroot-dir= --with-dnstap-socket-path=/run/dnstap.sock --disable-rpath --with-pidfile=/run/unbound.pid --with-libevent --enable-tfo-client --with-rootkey-file=/usr/share/dns/root.key --disable-flto --enable-tfo-server
Linked libs: libevent 2.1.12-stable (it uses epoll), OpenSSL 3.0.14 4 Jun 2024
Linked modules: dns64 python subnetcache respip validator iterator
TCP Fastopen feature available

BSD licensed, see LICENSE in source package for details.
Report bugs to unbound-bugs@nlnetlabs.nl or https://github.com/NLnetLabs/unbound/issues_

**Reply by Dynamic5912** [reply]
> And it does in fact compile with `--enable-cachedb --with-libhiredis`: https://buildd.debian.org/status/fetch.php?pkg=unbound&arch=amd64&ver=1.20.0-1%2Bb1&stamp=1719260287&raw=0
> 
> 
> 
> And here is the manpage for the Bookworm package, which contains all instructions to configure Unbound for using Redis as cachdb backaned: https://manpages.debian.org/unbound.conf

That package is for version **1.20.0-1+b1** whereas the latest release available on **Stable Debian** is **1.17.1**

**Reply by MichaIng** [reply]
Oh you are right. The manpage also says "if it was compiled with" ... The build log above is from Sid. The Bookworm build does indeed not contain it, and checking the changelog, it was added with package version 1.18.0-2, hence one of the earlier Sid builds, now in Trixie for upcoming Debian release: https://metadata.ftp-master.debian.org/changelogs//main/u/unbound/unbound_1.20.0-1_changelog

> unbound (1.18.0-2) unstable; urgency=medium
> 
>    * d/resolvconf-forwards: remove -e (Closes: #1035800), shorten sed expr
>    * d/changelog: mention #1013957 in previous changelog entry
>    * d/control, d/rules: switch from libnettle back to libssl once it is
>      GPL-compatible (#828699 is of no concern anymore).  This fixes libunbound
>      init failure.  Also Closes: #1007260
>    * d/control, d/rules: build daemon with --enable-cachedb --with-libhiredis,
>      build-depend on libhiredis-dev (Closes: #1014456)

There was a `bullseye-backports` version, sadly no `bookworm-backports`. But it would not hurt to ask the Debian maintainer for a backport of the current version shipped with Trixie. Can be done via bug tracker: https://bugs.debian.org/cgi-bin/pkgreport.cgi?src=unbound

**Reply by Dynamic5912** [reply]
How can I manually (or otherwise) upgrade Unbound to the latest release?

And would upgrading also add the additional components?

I note that Unbound provides info on downloading and compiling your own installer for the latest release - but I'm not savvy/comfortable enough to do that myself..

**Reply by Joulinar** [reply]
> How can I manually (or otherwise) upgrade Unbound to the latest release?

I guess that would mean switching from Debian Bookworm to the bleeding-edge Debian Trixi.

**Reply by MichaIng** [reply]
You would need to compile it yourself from source then. These instructions for Ubuntu should work on Debian as well: https://unbound.docs.nlnetlabs.nl/en/latest/getting-started/installation.html#id1

An alternative would be to upgrade the whole distribution to Debian Trixie. But it is still very volatile, might go through further larger package transitions, which by times means to manually resolve conflicts, remove obsolete packages and install their newer counter parts (like removing PHP 8.2 once PHP 8.3 is available in the repo and such). Security-wise and regarding bugs, Debian Trixie/testing is actually quite stable, comparable to any other rolling-release distribution or the non-LTS Ubuntu versions, but amount and kind of package upgrades requires more attention (also like on rolling-release distros like Arch Linux), hence some experience or mood to tinker 😉.

**Reply by Dynamic5912** [reply]
Thanks I'll give that a try. 

Do you know if installing/updating Unbound in-situ over an existing installation will change any existing configurations or override .conf files?

**Reply by MichaIng** [reply]
The configuration files should remain. Ours is anyway installed to `/etc/unbound/unbound.conf.d/dietpi.conf`, hence no pach where any default config could be installed to. The only thing I am not entirely sure about is whether the `sudo make install` step installs a new systemd unit over the existing one, respectively to `/etc/systemd/system/unbound.service` to use the new binary or not. The DEB package installation remains mostly untouched with e.g. `/usr/bin/unbound` while `make install` installs it to `/usr/local/bin/unbound`, where it takes priority. But the systemd service defines the absolute executable path.

You can check whether `/usr/local/bin/unbound` is used via e.g.
```sh
systemctl cat unbound
```
after compiling+install has finished.
